### PR TITLE
CB-12863 Introduce Distrox API change for supporting multiAZ. Availab…

### PR DESCRIPTION
**Pull request by doktoric**
…ility zones now presented on instancegroup level and availability zones as well. Also added the domain migration

See detailed description in the commit message.

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToTemplatePreparationObjectConverter.java [hunk 1]
```diff
@@ -297,8 +297,7 @@ private void decorateBuilderWithPlacement(StackV4Request source, Builder builder
         PlacementSettingsV4Request placementSettings = source.getPlacement();
         if (placementSettings != null) {
             String region = placementSettings.getRegion();
-            String availabilityZone = placementSettings.getAvailabilityZone();
-            builder.withPlacementView(new PlacementView(region, availabilityZone));
+            builder.withPlacementView(new PlacementView(region, region));
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToTemplatePreparationObjectConverter.java:300]: why did you remove the az field?

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/instancegroup/network/InstanceGroupNetworkV4ToInstanceGroupNetworkConverter.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.sequenceiq.cloudbreak.domain.stack.instance.network.InstanceGroupNetwork;
+
+@Component
+public class InstanceGroupNetworkV4ToInstanceGroupNetworkConverter
+    extends AbstractConversionServiceAwareConverter<InstanceGroupNetworkV4Request, InstanceGroupNetwork> {
+
+    @Override
+    public InstanceGroupNetwork convert(InstanceGroupNetworkV4Request source) {
+        InstanceGroupNetwork network = new InstanceGroupNetwork();
+        if (source.getAws() != null) {
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/instancegroup/network/InstanceGroupNetworkV4ToInstanceGroupNetworkConverter.java:21]: why don't you use the `ProviderParameterCalculator providerParameterCalculator;`?

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkV1ToInstanceGroupNetworkV4Converter.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        if (key == null) {
+            key = new InstanceGroupNetworkV1Request();
+        }
+
+        InstanceGroupNetworkV4Request request = new InstanceGroupNetworkV4Request();
+
+        if (network.getValue() != null) {
+            switch (value.getCloudPlatform()) {
+                case "AWS":
+                    request.setAws(getAwsNetworkParameters(Optional.ofNullable(key.getAws()), environmentNetworkResponse));
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkV1ToInstanceGroupNetworkV4Converter.java:47]: why don't you use the `ProviderParameterCalculator providerParameterCalculator;`?

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToTemplatePreparationObjectConverter.java (hunk 1 above)
Comment (doktoric) [reply core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToTemplatePreparationObjectConverter.java:300]: this field is not even used in the template generation. Also we are nor always populated this value

Comment (topolyai5) [reply core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToTemplatePreparationObjectConverter.java:300]: If I understand correctly, we wouldn't set the az via API?

File: core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+-- Search subnet-id
+UPDATE instancegroupnetwork SET subnetid=instancegroupnetwork.attributes::json->'subnetId';
+UPDATE instancegroupnetwork SET subnetid = REPLACE(subnetid,'"','');
+
+-- If AWS, AZURE, GCP, OPENSTACK, MOCK
+UPDATE instancegroupnetwork SET attributes = json_build_object('subnetIds', json_build_array(subnetid), 'cloudPlatform', cloudPlatform) WHERE subnetid IS NOT NULL;
+
+-- If YCLOUD
+UPDATE instancegroupnetwork SET attributes = json_build_object('subnetIds', json_build_array(), 'cloudPlatform', cloudPlatform) WHERE subnetid IS NULL;
+UPDATE instancegroup ig SET instancegroupnetwork_id=ign.instancegroup_id FROM instancegroupnetwork ign WHERE ign.instancegroup_id = ig.id;
```
Comment (lajosrodek) [inline core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql:36]: >`instancegroupnetwork_id=ign.instancegroup_id`

Here I think `ign.instancegroup_id` should be replaced with `ign.id`. Furthermore, this statement probably applies to all providers, not only YCloud, so one or more empty lines should be added before.

File: core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+-- Cleaning up tmp columns
+ALTER TABLE instancegroupnetwork DROP COLUMN stack_network_id;
+ALTER TABLE instancegroupnetwork DROP COLUMN subnetid;
+ALTER TABLE instancegroupnetwork DROP COLUMN IF EXISTS instancegroup_id;
```
Comment (lajosrodek) [inline core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql:41]: Not a big deal but these could be collapsed into a single `ALTER TABLE` statement.

File: core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql [hunk 6]
```diff
@@ -0,0 +1,49 @@
+-- // CB-1068 multi az
+-- Migration SQL that makes the change goes here.
+
+CREATE TABLE IF NOT EXISTS instancegroupnetwork (
+    id bigint NOT NULL,
+    cloudPlatform CHARACTER VARYING(255),
+    attributes TEXT,
+    stack_network_id bigint,
+    instancegroup_id bigint,
+    subnetid TEXT,
```
Comment (lajosrodek) [inline core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql:10]: I know this is only a tmp column. `instancemetadata.subnetid` is a `VARCHAR(255)`, so perhaps the same type could be sufficient here too.

File: core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql (hunk 5 above)
Comment (lajosrodek) [inline core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql:41]: How about dropping `CONSTRAINT fk_instancegroup_network_id` on `instancegroupnetwork` explicitly before removing the column `instancegroup_id`?

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupRequestToInstanceGroupConverterTest.java [hunk 7]
```diff
@@ -28,6 +29,7 @@
 import com.sequenceiq.cloudbreak.domain.SecurityGroup;
 import com.sequenceiq.cloudbreak.domain.Template;
 import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceGroup;
+import com.sequenceiq.cloudbreak.domain.stack.instance.network.InstanceGroupNetwork;
 
 @RunWith(MockitoJUnitRunner.class)
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupRequestToInstanceGroupConverterTest.java:34]: Optional: Migrate to JUnit 5.

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupToInstanceGroupResponseConverterTest.java [hunk 8]
```diff
@@ -48,14 +49,15 @@ public void testConvert() {
 
         assertEquals(source.getNodeCount(), result.getNodeCount());
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupToInstanceGroupResponseConverterTest.java:50]: Optional: Migrate to JUnit 5 (applies to whole class).

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupToInstanceGroupResponseConverterTest.java [hunk 9]
```diff
@@ -48,14 +49,15 @@ public void testConvert() {
...
+        assertAllFieldsNotNull(result, Lists.newArrayList("recipes", "template", "network",
+                "securityGroup", "gcp", "name", "mock", "openstack",
                 "aws", "yarn", "azure"));
     }
 
     @Override
     public InstanceGroup createSource() {
-        var instanceGroup = TestUtil.instanceGroup(1L, InstanceGroupType.CORE, TestUtil.gcpTemplate(1L));
-        var metaData = TestUtil.instanceMetaData(1L, 1L, InstanceStatus.SERVICES_RUNNING, false, instanceGroup);
+        InstanceGroup instanceGroup = TestUtil.instanceGroup(1L, InstanceGroupType.CORE, TestUtil.gcpTemplate(1L));
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupToInstanceGroupResponseConverterTest.java:59]: I support eliminating `var` in favor of explicit types, so perhaps you could do the same to the remaining `var` occurrences as well. 😄

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/v2/StackV4RequestToTemplatePreparationObjectConverterTest.java [hunk 10]
```diff
@@ -437,7 +437,7 @@ public void testStackPlacement() {
         TemplatePreparationObject result = underTest.convert(source);
         assertTrue(result.getPlacementView().isPresent());
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/converter/v2/StackV4RequestToTemplatePreparationObjectConverterTest.java:438]: Optional: Migrate to JUnit 5 (applies to whole class).

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupRequestToInstanceGroupConverterTest.java [hunk 11]
```diff
@@ -28,6 +29,7 @@
...
 public class InstanceGroupRequestToInstanceGroupConverterTest extends AbstractJsonConverterTest<InstanceGroupV4Request> {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/converter/stack/instance/InstanceGroupRequestToInstanceGroupConverterTest.java:35]: Rename class: Insert `V4` before `Request`.

File: core/src/test/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4ConverterTest.java [hunk 12]
```diff
@@ -104,6 +117,22 @@ void convertToAwsWithEnvironmentWithoutSecurityGroupsHappyPathTest() {
         assertThat(first.getType()).isEqualTo(InstanceGroupType.CORE);
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4ConverterTest.java:117]: What about checking for `subnetId` (`"aws-123"`) or the whole network block?

File: core/src/main/java/com/sequenceiq/cloudbreak/service/location/LocationProvider.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package com.sequenceiq.cloudbreak.service.location;
+
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+
+@Service
+public class LocationProvider {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/location/LocationProvider.java:8]: Cool, what is this required for? 😄

File: core/src/main/java/com/sequenceiq/cloudbreak/service/decorator/StackDecorator.java [hunk 14]
```diff
@@ -220,11 +227,38 @@ private String getAvailabilityZoneFromEnv(DetailedEnvironmentResponse environmen
         }
     }
 
+    private Set<String> getAvailabilityZoneFromEnv(InstanceGroup group, DetailedEnvironmentResponse environment) {
+        if (group.getInstanceGroupNetwork() != null && environment.getNetwork() != null
+                && environment.getNetwork().getSubnetMetas() != null
+                && (group.getAvailabilityZones() == null || group.getAvailabilityZones().isEmpty())) {
+            return getAvailabilityZones(environment, group.getInstanceGroupNetwork().getAttributes());
+        } else {
+            return Set.of();
```
[the remark this answers is not in the document]
Comment (doktoric) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/decorator/StackDecorator.java:236]: mock tests does not have availability zone

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java [hunk 15]
```diff
@@ -141,4 +160,117 @@ private void setSecurityAccess(SecurityGroupV4Request securityGroup, String secu
...
+            case "GCP":
+                validateSubnet(network, environment, network.getGcp().getSubnetIds());
+                break;
+            default:
+        }
+    }
+
+    private void validateSubnet(InstanceGroupNetworkV4Request network, DetailedEnvironmentResponse environment, List<String> subnetIds) {
+        if (subnetIds != null && !subnetIds.isEmpty()
+                && (environment == null
```
[the remark this answers is not in the document]
Comment (doktoric) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java:267]: not really. the old stack api still works

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java [hunk 16]
```diff
@@ -141,4 +160,117 @@ private void setSecurityAccess(SecurityGroupV4Request securityGroup, String secu
...
+            case "AWS":
+                validateSubnet(network, environment, network.getAws().getSubnetIds());
+                break;
+            case "AZURE":
+                validateSubnet(network, environment, network.getAzure().getSubnetIds());
+                break;
+            case "GCP":
+                validateSubnet(network, environment, network.getGcp().getSubnetIds());
+                break;
+            default:
```
[the remark this answers is not in the document]
Comment (doktoric) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java:261]: no need to validate

File: core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql (hunk 4 above)
Comment (doktoric) [reply core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql:36]: i think the statement is valid

File: core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql (hunk 6 above)
Comment (doktoric) [reply core/src/main/resources/schema/app/20200824205107_CB-1068_multi_az.sql:10]: is this matter?

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+    private InstanceGroupAwsNetworkV4Parameters convertToAwsNetworkRequest(Pair<InstanceGroupAwsNetworkV1Parameters,
+            EnvironmentNetworkResponse> source) {
+        InstanceGroupAwsNetworkV1Parameters key = source.getKey();
+
+        InstanceGroupAwsNetworkV4Parameters response = new InstanceGroupAwsNetworkV4Parameters();
+
+        if (key != null) {
+            List<String> subnetIds = key.getSubnetIds();
+            if (subnetIdsNotDefined(subnetIds)) {
+                response.setSubnetIds(key.getSubnetIds());
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java:164]: Nit: `key.getSubnetIds()` can be replaced with `subnetIds`.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+    private InstanceGroupOpenstackNetworkV4Parameters convertToOpenstackNetworkRequest(Pair<InstanceGroupOpenstackNetworkV1Parameters,
+            EnvironmentNetworkResponse> source) {
+        InstanceGroupOpenstackNetworkV1Parameters key = source.getKey();
+
+        InstanceGroupOpenstackNetworkV4Parameters response = new InstanceGroupOpenstackNetworkV4Parameters();
+
+        if (key != null) {
+            List<String> subnetIds = key.getSubnetIds();
+            if (subnetIdsNotDefined(subnetIds)) {
+                response.setSubnetIds(key.getSubnetIds());
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java:182]: Nit: `key.getSubnetIds()` can be replaced with `subnetIds`.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+                response.setSubnetIds(List.of(source.getValue().getPreferedSubnetId()));
+            }
+        }
+
+        return response;
+    }
+
+    private boolean subnetIdsNotDefined(List<String> subnetIds) {
+        return subnetIds != null && !subnetIds.isEmpty();
+    }
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java:193]: The method name is confusing. It returns `true` if there are some subnet IDs in the list.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        InstanceGroupMockNetworkV1Parameters key = source.getKey();
+
+        InstanceGroupMockNetworkV4Parameters params = new InstanceGroupMockNetworkV4Parameters();
+
+        if (key != null) {
+            List<String> subnetIds = key.getSubnetIds();
+            if (value != null) {
+                if (subnetIdsNotDefined(subnetIds)) {
+                    params.setSubnetIds(subnetIds);
+                } else {
```
[the remark this answers is not in the document]
Comment (doktoric) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupNetworkParameterConverter.java:105]: yes

**Review by lajosrodek — APPROVED** [review]
Thanks for the fixes, LGTM 🎉
